### PR TITLE
post: More correct 'wait' logic

### PR DESCRIPTION
**Pull request by magik6k**

### CODE REVIEW

File: chain/deals/client.go [hunk 1]
```diff
@@ -84,7 +84,7 @@ func NewClient(sm *stmgr.StateManager, chain *store.ChainStore, h host.Host, w *
 		dag:          dag,
 		discovery:    discovery,
 		fm:           fm,
-		events:       events.NewEvents(context.TODO(), &chainapi),
+		events:       events.NewEvents(context.TODO(), &struct {full.ChainAPI; full.StateAPI}{chainapi, stateapi}),
```
Comment (whyrusleeping) [inline chain/deals/client.go]: no

File: api/api_full.go [hunk 2]
```diff
@@ -110,6 +110,7 @@ type FullNode interface {
 	StateMarketParticipants(context.Context, *types.TipSet) (map[string]actors.StorageParticipantBalance, error)
 	StateMarketDeals(context.Context, *types.TipSet) (map[string]actors.OnChainDeal, error)
 	StateMarketStorageDeal(context.Context, uint64, *types.TipSet) (*actors.OnChainDeal, error)
+	StateGetReceipt(context.Context, cid.Cid, *types.TipSet) (*types.MessageReceipt, error)
```
Comment (whyrusleeping) [inline api/api_full.go]: I think this is a duplicate of `StateReplay`

Comment (magik6k) [reply api/api_full.go]: This one looks back in chain.

File: chain/events/utils.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+		more, err = hnd(msg, rec, ts, ts.Height())
+
+		return done, more, err
+	}
+}
+
+func(e *calledEvents) MatchMsg(inmsg *types.Message) MatchFunc {
+	return func(msg *types.Message) (bool, error) {
+		return inmsg.Equals(msg), nil
```
Comment (whyrusleeping) [inline chain/events/utils.go]: should add:
```go
if msg.From == inmsg.From && msg.Nonce == inmsg.Nonce && !inmsg.Equals(msg) {
    return false, ErrWhatDoWeDoNow
}
```

File: storage/miner.go [hunk 4]
```diff
@@ -48,17 +48,16 @@ type Miner struct {
...
-
 	// Call a read only method on actors (no interaction with the chain required)
 	StateCall(ctx context.Context, msg *types.Message, ts *types.TipSet) (*types.MessageReceipt, error)
 	StateMinerWorker(context.Context, address.Address, *types.TipSet) (address.Address, error)
 	StateMinerProvingPeriodEnd(context.Context, address.Address, *types.TipSet) (uint64, error)
 	StateMinerSectors(context.Context, address.Address, *types.TipSet) ([]*api.ChainSectorInfo, error)
 	StateMinerProvingSet(context.Context, address.Address, *types.TipSet) ([]*api.ChainSectorInfo, error)
 	StateMinerSectorSize(context.Context, address.Address, *types.TipSet) (uint64, error)
-	StateWaitMsg(context.Context, cid.Cid) (*api.MsgWait, error)
+	StateWaitMsg(context.Context, cid.Cid) (*api.MsgWait, error)  // TODO: removeme
```
Comment (whyrusleeping) [inline storage/miner.go]: we would still need this exposed through the api somehow

**Comment by magik6k** [comment]
(rebased)

**Pull request merged**